### ISSUE TITLE
[Bug] Capture is not working for Mozilla Firefox at the moment

### ISSUE THREAD

**Issue by CrendKing** [jc3213/download_with_aria2-chromium#1]
Unlike the [Chrome version](https://developer.chrome.com/extensions/downloads#type-DownloadItem), the [DownloadItem API in Firefox](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/downloads/DownloadItem) has no `finalUrl` property. Therefore [`captureCheck()` and `captureAdd()`](https://github.com/jc3213/download_with_aria2/blob/master/background.js#L47) are incorrect. aria2c will always complain that URI is not provided.

However, even after I fix that problem locally, the "Capture" function is not working as intended. What I expect is that the original download request is intercepted, and only the aria2 download proceeds. What actually happening is that I'm getting double downloads, one from Firefox and the other from aria2.

I think in order to intercept the download, you need to use the [webRequest API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest), much as what [this is doing](https://github.com/RossWang/Aria2-Integration/blob/a45b57d88283bb4e85e53c8be031810d0eddb993/App/common.js). At the moment, only the context menu item is useful.

BTW, in Firefox, the extension APIs are in `browser` namespace, as opposed to `chrome` in Chrome, even though Firefox compatibility layer [accepts both](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities). You should consider switching to the more idiomatic `browser` if you don't have intention to port this back to Chrome (considering this is originally ported from Chrome).

**Comment by jc3213** [comment]
Thanks for the report. I will try to investigate this matter, but I can't promise you when it will be fixed. I just coding better than a new programmer.

>However, even after I fix that problem locally, the "Capture" function is not working as intended. What I expect is that the original download request is intercepted, and only the aria2 download proceeds. What actually happening is that I'm getting double downloads, one from Firefox and the other from aria2.
> 
If `captureAdd()` is called, it will get will trigger `chrome.downloads.erase` to remove the download. This works fine on Chromium. maybe another firefox compatibility problem.

PS: 
I used to be a heavy firefox user, but now I'm sticking on Chromium. I hate the new Firefox altogether with its memory leaks, and I'm still working on the popup window.

I uesd to complain that Mozilla has made a 70% Chromium + alpha, while Microsoft makes a 40% Chromium + beta. Then Microsoft has switched to Chomium. and Mozilla still not supporting many useful apis.

**Comment by jc3213** [comment]
I've looked into the code you've provided. and tried to figure it out on Vivaldi. I was shocked by the number of logs in console. I'd fairly say that using `webRequest` is a very slow, and very complicate solution to this matter.

It must judge firstly the `Content-Type` to filter which response `details` is needed from many other requests. Then filter `Content-Disposition` to retrieve filename, then matching through all `Capture Filters`. These are done to every single `webRequest` call, even the `jsonrpc` call from `popup.html`

It will never be a solution but a `better than none` work around to me. BTW, the author has written a url filter also, then his add-on won't support magnet either. this just like many forks of the former `aria2c-integration`.

**Comment by CrendKing** [comment]
> If `captureAdd()` is called, it will trigger `chrome.downloads.erase` to remove the download. This works fine on Chromium. maybe another firefox compatibility problem. console logs will help

Here is the description of `[chrome.downloads.erase](https://developer.chrome.com/extensions/downloads#method-erase)` from Chrome:

> Erase matching DownloadItem from history without deleting the downloaded file.

This description does not mention about either 1) stop the current download; or 2) delete any file, whether it is still downloading or finished download. The event order I observe in Firefox is:
1. User click link.
2. Firefox prompt download dialog, choose save path.
3. Firefox download starts.
4. `browser.download.onCreated` triggers.
5. This extension deletes the download history, and send request to aria2.

Unfortunately, even after step 5, Firefox is still downloading the file. You can't remove the temporary ".part" file (which is eventually renamed to the save name after completion). This is what I meant "double download". The file is actually downloaded twice.

>  using webRequest is a very slow, and very complicate solution to this matter.

It could be complicated, but I'm not sure about the "slow" part. I use uBlock Origin all the time, and never felt it slows the browser.

The way to only intercept download requests requires inspecting headers like `Content-Disposition`. However, it could trigger false positives with interactions with other extensions. That's why I'm looking for an alternative of aria2 integration. I feel as long as the context menu is functioning properly I'll stick with your extension. I don't really need the capture, and I'll only use aria2 to download large files anyways. But if you strive to perfect everything, I hope this issue is something you could look back.

**Comment by jc3213** [comment]
>If captureAdd() is called, it will trigger chrome.downloads.erase to remove the download. This works fine on Chromium. maybe another firefox compatibility problem. console logs will help

Here is the description of [chrome.downloads.erase](https://developer.chrome.com/extensions/downloads#method-erase) from Chrome:

Erase matching DownloadItem from history without deleting the downloaded file.

This description does not mention about either 1) stop the current download; or 2) delete any file, whether it is still downloading or finished download. The event order I observe in Firefox is:

User click link.
Firefox prompt download dialog, choose save path.
Firefox download starts.
browser.download.onCreated triggers.
This extension deletes the download history, and send request to aria2.
Unfortunately, even after step 5, Firefox is still downloading the file. You can't remove the temporary ".part" file (which is eventually renamed to the save name after completion). This is what I meant "double download". The file is actually downloaded twice.
>
Both `onCreated` or `onDeterminingFileName` event handler won't download anything or leave a stopped download in history on Chrome. But `downloads.cancel` method will leave a stopped download 0/filesize in history.

>It could be complicated, but I'm not sure about the "slow" part. I use uBlock Origin all the time, and never felt it slows the browser.
>
It is because your CPU is powerful enough. an old laptop CPU, or something like Atom will suffer a lot

>The way to only intercept download requests requires inspecting headers like Content-Disposition. However, it could trigger false positives with interactions with other extensions. That's why I'm looking for an alternative of aria2 integration. I feel as long as the context menu is functioning properly I'll stick with your extension. I don't really need the capture, and I'll only use aria2 to download large files anyways. But if you strive to perfect everything, I hope this issue is something you could look back.
>
Yes, that's true. and Chromium provides `downloads.onDeterminingFileName` event handler, that privides almost all info that are needed.
I've once tried to rewrite my former XUL extension (which uses NetUtils.jsm) with webRequest and find that I can't add more than one event handler on Firefox. That's why my old extensions are all out-of-dated and I have removed all of them from Github and AMO.

Maybe I should provide a firefox extension without `Capture` option, untill a better solution is out.

**Comment by jc3213** [comment]
After this extension is done, I will start a new repository for Firefox port with `webRequest`.

`Capture` function will be changed to match `File Size` or `File Extension` together with `Ignored Hosts`. the `Monitored Hosts` option will be deprecated because of `webRequest`.

**Comment by CrendKing** [comment]
The monitored hosts could be filter by the extension logic (like do an `if` inside the event listener), no? I don't mind deprecating it, but someone else might find it useful.

**Comment by jc3213** [comment]
Because `webRequest` doesn't provide properties as simple as `downloads.downloadItem`, matching `Monitored Hosts`  will need to match `MIME` types or other properties first, in order that not `<all-urls>` from that host are monitored. 

I've once tried the `MIME` types but it is very complicated, and confusing. For example. MIME of `.exe` is `application/vnd.microsoft.portable-executable` according to (https://www.iana.org/assignments/media-types/media-types.xhtml). But browser returns `application/octet-stream`.

If you write a short list. you must use either `regexp.test(mime)` or `array.prototype.filter(item => mime.includes(item))` first, then matching file extension, url, filesize.

It's buggy too, for exapme if you monitored `www.google.com`, with `jpg` file extension. All files with `jpg` file extension from host `www.google.com` will be captured, every time you open a new www.google.com link.

**Comment by jc3213** [comment]
Have a try: https://github.com/jc3213/download_with_aria2-firefox

**Comment by CrendKing** [comment]
> Have a try: https://github.com/jc3213/download_with_aria2-firefox

The context menu is working. The capture has no effect (tried one .exe link and one .7z link).

**Comment by jc3213** [comment]
Well, Its really buggy, I downloads `*.html` instead of open the website. I just monitored `exe` file extension. WTF.

I'm trying to fix it, but since I haven't using Firefox for years, it will be very a long term though.

**Comment by jc3213** [comment]
> `browser.downloads.onCreated` in Firefox, just work as `chrome.downloads.onDeterminingFilename` in Chrome.
Both are triggered after prompting the window to ask the filename, and the properties are almost the same (Only Firefox doens't provide `item.finalUrl`)
>
That's the how `downloads` api works
> If extracting file extension from `details.url` is wrong, it returns '' .and will be always matching fileExt option. so Firefox capture `www.example.com` and send it to Aria2
>>https://github.com/jc3213/download_with_aria2-firefox/blob/5f81c706f8948e9f27e1b9f5cacc011bb2ea2098/background.js#L35
>>
>
Well. It's better to try `downloads` api first though

**Comment by jc3213** [comment]
>Unfortunately, even after step 5, Firefox is still downloading the file. You can't remove the temporary ".part" file (which is eventually renamed to the save name after completion). This is what I meant "double download". The file is actually downloaded twice.
>
I've found why this happens. Firefox handles `downloads.erase(id)` while chromium handles `downloads.erase( {id: id} )`, that's why Firefox doesn't stop downloads or remove it from history.

Well, I'd say that's shitty bitty

**Comment by jc3213** [comment]
Here's the signed xpi, rename `.xpi.zip` to `.xpi`, since github doesn't support upload `.xpi`
[download_with_aria2-2.1100-fx.xpi.zip](https://github.com/jc3213/download_with_aria2-chromium/files/4795824/download_with_aria2-2.1100-fx.xpi.zip)

**Comment by jc3213** [comment]
AMO https://addons.mozilla.org/en-US/firefox/addon/downwitharia2/

**Comment by CrendKing** [comment]
Hi. Just got a notification about this ticket months ago. I thought the Firefox API does not allow capture right? I just tried it. Here's result:

* "File Extensions" I put `exe`. Then proceed to download a .exe file. The normal Firefox download dialog pops up, ask me where to save. Then "Downloads" window from Library shows me "File is moved or missing". It is as if Firefox started download, then something forcefully stopped it, but never took over the job. I checked the addon and there was nothing in it.

* "File Extensions" put `*`. Then proceed to download .exe. Everything happened as the stock Firefox. Nothing from addon.

**Comment by jc3213** [comment]
@CrendKing I'll look into this problem, thanks for your report.

- I wonder which revision you are using?
- The downloads from Firefox is stopped by `downloads.cancel()` because `downloads.remove()` only remove it from history but won't stop downloads as you've reported.

**Comment by jc3213** [comment]
@CrendKing I've located the bug. this is because of once I've changed the code structure and just copy and paste from chromium code.
in background.js line 32 `downWithAria2(item.finalUrl, item.referrer);` should be `downWithAria2(item.url, item.referrer);`
It's a simple mistake that has been fixed once though. sorry for the inconvenience

**Comment by CrendKing** [comment]
> downWithAria2(item.finalUrl, item.referrer); should be downWithAria2(item.url, item.referrer);

Oh, that one :)

Wow, I can't believe you actually get it working! It is capturing and intercepting the download! Could you be THE first person in the whole world that achieve this with WebExtension Firefox, because all other popular download addons (such as DownThemAll) are still doing the context menu approach.

OK, may I ask for some feature requests?

1. Can I put "*" in the "File Extensions" so it captures all downloads?
2. Could you send the chosen download directory to aria2? Currently if I launch aria2 without `--dir`, all files are downloaded to the same directory as aria2c.exe. If I specify `--dir`, all files go there, regardless where I choose in the Firefox popup dialog. I think since you can access the Firefox download object, you must be able to know where the file is saved, and aria2 RPC interface allows you to pass directory.
3. If 2 is implemented, maybe update the addon popup so that it shows the directory as well as the filename. Currently it only shows filename.

Thank you.

**Comment by jc3213** [comment]
This is the answer
1. No, the whildcard `*` is not supported. If you'd like always capture the downloads, there should be some changes on the `options.html` for `Capture` settings.
2. I've once tried to do it, but I'm failed on making ui with `<input type="file">`, Use only string is some how a bit inconvenient
3. This can be implemented so, but since the size of the popup window is limited, I don't think display full path of the file is a good choice

**Comment by CrendKing** [comment]
1. Do you mean you'd prefer to add a separate checkbox saying "Monitor all downloads" instead of * in the 3 boxes? I'm OK either case, as long as it is all downloads.
2. I don't see why you mention UI change. The `DownloadItem.filename` is absolute path. You can split the directory name, and pass "--dir" to the RPC option. Everything is opaque to end user. Or you have some reason disliking it?
3. This one is not very important. If you are afraid of the width, you can just put the full path to an `alt` tooltip.

**Comment by jc3213** [comment]
1. This has been done though, `Capture` is now an option menu not a checkbox. three values. `Disabled`, `Built-in Filters`, `Always`, I'm testing it on Chromium and worked perfectly. I'll merge the codes in to firefox side, but some bug might happen(because of copy paste)
2. I seemed to misunderstand your meaning? You are asking an option to download files into where browser saves them??
3. `alt` tooltip might be a choice. I'll try later though.

**Comment by CrendKing** [comment]
1. Thanks.
2. Yes. For example, hit a link, Firefox asks "Open with", "Save File". Choose "Save File", it asks where to save to. If I capture, the download will be cancelled and taken over by the addon. I want the captured file to save to where I chose.
3. Thanks.

**Comment by jc3213** [comment]
1. I've pushed new version to AMO, please wait
2. OK, i got it. But it seems this can only be done for firefox though? Then this should be the first Firefox exclusive feature.

Your request is what Chrome version have. `Capture Downloads` based on `Built-in Filters` then `File Size` -> `1` `MB` should work. Firefox only returns unknown size (-1).
To retrieving filesize, it needs to call XMLHttpRequest or fetch(I'm don't know much about fetch) that might have CORS problems.

**Comment by CrendKing** [comment]
For 2, why is it Firefox only? Chrome's `DownloadItem.filename` is also described as "Absolute local path", so I assume you can split the parent directory out, right?

For Firefox unknown size, that's definitely not intended behavior, if Chrome's same API works. Maybe we should create a ticket in Bugzilla. Could take them years to fix though, considering their lack of resource right now.

I could try implementing 2 and send a PR tomorrow.

**Comment by jc3213** [comment]
UI for 2, but no js, or css, don't know whether this is good enough?
![1](https://user-images.githubusercontent.com/8744936/93710576-6b1f0b80-fb7a-11ea-88c0-6ec68d65f419.png)

**Comment by CrendKing** [comment]
PR: https://github.com/jc3213/download_with_aria2-firefox/pull/1

I got what you mean. There are 3 ways to add a download, the my PR only support the capture case. Your picture could solve the context menu option. For the manual add, you might need to add additional UI components.

However, if you want to support this in full fledge, I think all you need to do is add a check box in the capture tab saying "pass through the download directory". Leave the other 2 use cases as before. User needs to specify `--dir` when launching their aria2 daemon though.

**Comment by jc3213** [comment]
@CrendKing Based on your PR, I've found a typo... I think it's better pass an `Object` rather than multiple `Variables`... I've thought to rewrite this part though....

**Comment by CrendKing** [comment]
Sure. That PR is a mock up. You do however you want.

BTW, Firefox bug I just filed: https://bugzilla.mozilla.org/show_bug.cgi?id=1666137 . Hope they can take a look.

**Comment by CrendKing** [comment]
One person replied the bug with some info. I'm not sure how useful would it be. I'm thinking, one way to do the conditional pass through is simply wait for a small amount of time (e.g. 1 second) after onCreate(). There are 3 possible outcomes after the wait:

1. The download is already completed. Then we are good. No need to pass. The file must be extremely small.
2. `totayBytes` becomes available, as the dev commented on that bug. Then do the conditional check.
3. `totalBytes` not available, and download still going. Always pass since the file must be large, relative to user's download speed.

**Comment by jc3213** [comment]
If downloads starts (for any period of time, even less than 1 second), there will be I/O on your storage device. I can hardly remember if Firefox acquires the whole file size on disk to avoid fragments or not. Anyway, this is not a good solution for traditional hard-disk users either.

I'm thinking of using `webRequest` altogether with `downloads.onCreated` but how to synchronize data between two different APIs? or try XHR which may meet CORS problem on some sites?

**Comment by jc3213** [comment]
Here's the resolution based on webRequest
[firefox-webrequest.zip](https://github.com/jc3213/download_with_aria2-chromium/files/5272968/firefox-webrequest.zip)

Please have a try

**Comment by jc3213** [comment]
I've tried XHR, the download must be `cancel()` first then use `XHR.getResponseHeader()` to retrieve `Content-Length`. do matching, then `resume()` if not match, or `erase()` if matched. I don't want to change the code structure because it's hard to maintain two different code styles.

The `webRequest` workaround uses `array.filter()` twice, and logging recent 15 requests with `Content-Length` in responseHeaders. work for me though. The codes will not be pushed to `master` branch though

**Comment by jc3213** [comment]
oops. I've looked into Firefox mechanism on downloading. `onCreated` triggers the same time while file is writing to your device. Holy shit if I must have to rewrite this part. dammed

**Comment by jc3213** [comment]
The work around above seems have some problem (downloads will start and save files to local disk then get cancelled), try the new one here.

PS: If there's any other question, start a new issue, since this is too long.

[firefox r2100 with webRequest.zip](https://github.com/jc3213/download_with_aria2-chromium/files/5273791/firefox.r2100.with.webRequest.zip)
